File: metadata/mill-metadata-core/src/test/java/io/qpointz/mill/metadata/repository/file/FileFacetTypeRepositoryTest.java
package io.qpointz.mill.metadata.repository.file;

import io.qpointz.mill.metadata.domain.FacetTypeDescriptor;
import io.qpointz.mill.metadata.domain.MetadataTargetType;
import org.junit.jupiter.api.Test;

import java.util.Collection;
import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

class FileFacetTypeRepositoryTest {

    private FileFacetTypeRepository createRepository() {
        return new FileFacetTypeRepository(
                List.of("classpath:metadata/facet-types-test.yml"),
                new ClasspathResourceResolver());
    }

    @Test
    void shouldLoadFacetTypes_fromYaml() {
        var repo = createRepository();
        Collection<FacetTypeDescriptor> all = repo.findAll();
        assertEquals(2, all.size());
    }

    @Test
    void shouldFindByTypeKey() {
        var repo = createRepository();
        var descriptive = repo.findByTypeKey("descriptive");
        assertTrue(descriptive.isPresent());
        assertEquals("Descriptive", descriptive.get().getDisplayName());
        assertTrue(descriptive.get().isMandatory());
    }

    @Test
    void shouldLoadApplicableTo() {
        var repo = createRepository();
        var descriptive = repo.findByTypeKey("descriptive").orElseThrow();
        Set<MetadataTargetType> targets = descriptive.getApplicableTo();
        assertNotNull(targets);
        assertTrue(targets.contains(MetadataTargetType.TABLE));
        assertTrue(targets.contains(MetadataTargetType.SCHEMA));
        assertTrue(targets.contains(MetadataTargetType.ATTRIBUTE));
    }

    @Test
    void shouldLoadContentSchema() {
        var repo = createRepository();
        var audit = repo.findByTypeKey("custom-audit").orElseThrow();
        assertTrue(audit.hasContentSchema());
        assertNotNull(audit.getContentSchema());
        assertEquals("object", audit.getContentSchema().get("type"));
    }

    @Test
    void shouldSaveNewType() {
        var repo = createRepository();
        repo.save(FacetTypeDescriptor.builder()
                .typeKey("new-type")
                .displayName("New Type")
                .build());
        assertTrue(repo.existsByTypeKey("new-type"));
    }

    @Test
    void shouldDeleteType() {
        var repo = createRepository();
        assertTrue(repo.existsByTypeKey("descriptive"));
        repo.deleteByTypeKey("descriptive");
        assertFalse(repo.existsByTypeKey("descriptive"));
    }

    @Test
    void shouldReturnEmpty_forUnknownType() {
        var repo = createRepository();
        assertTrue(repo.findByTypeKey("nonexistent").isEmpty());
    }
}

File: metadata/mill-metadata-core/src/test/kotlin/io/qpointz/mill/metadata/repository/file/FileFacetTypeRepositoryTest.kt
package io.qpointz.mill.metadata.repository.file

import io.qpointz.mill.metadata.domain.FacetTypeDescriptor
import io.qpointz.mill.metadata.domain.MetadataUrns
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class FileFacetTypeRepositoryTest {

    private fun createRepository(): FileFacetTypeRepository =
        FileFacetTypeRepository(listOf("classpath:metadata/facet-types-test.yml"), ClasspathResourceResolver())

    @Test fun shouldLoadFacetTypes_fromYaml() {
        assertEquals(2, createRepository().findAll().size)
    }

    @Test fun shouldFindByTypeKey() {
        val descriptive = createRepository().findByTypeKey(MetadataUrns.FACET_TYPE_DESCRIPTIVE)
        assertTrue(descriptive.isPresent)
        assertEquals("Descriptive", descriptive.get().displayName)
        assertTrue(descriptive.get().mandatory)
    }

    @Test fun shouldLoadApplicableTo() {
        val targets = createRepository().findByTypeKey(MetadataUrns.FACET_TYPE_DESCRIPTIVE)
            .orElseThrow().applicableTo
        assertNotNull(targets)
        assertTrue(MetadataUrns.ENTITY_TYPE_TABLE in targets!!)
        assertTrue(MetadataUrns.ENTITY_TYPE_SCHEMA in targets)
        assertTrue(MetadataUrns.ENTITY_TYPE_ATTRIBUTE in targets)
    }

    @Test fun shouldLoadContentSchema() {
        val audit = createRepository().findByTypeKey("custom-audit").orElseThrow()
        assertTrue(audit.hasContentSchema())
        assertNotNull(audit.contentSchema)
        assertEquals("object", audit.contentSchema!!["type"])
    }

    @Test fun shouldSaveNewType() {
        val repo = createRepository()
        repo.save(FacetTypeDescriptor(typeKey = "new-type", displayName = "New Type"))
        assertTrue(repo.existsByTypeKey("new-type"))
    }

    @Test fun shouldDeleteType() {
        val repo = createRepository()
        assertTrue(repo.existsByTypeKey(MetadataUrns.FACET_TYPE_DESCRIPTIVE))
        repo.deleteByTypeKey(MetadataUrns.FACET_TYPE_DESCRIPTIVE)
        assertFalse(repo.existsByTypeKey(MetadataUrns.FACET_TYPE_DESCRIPTIVE))
    }

    @Test fun shouldReturnEmpty_forUnknownType() {
        assertTrue(createRepository().findByTypeKey("nonexistent").isEmpty)
    }
}
